# Importación de Librerías

En esta sección se importan las librerías necesarias para el análisis y limpieza de datos.

In [156]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

pd.options.future.infer_string = False

## Carga y Exploración Inicial

Se carga el dataset y se realiza una inspección inicial.

In [119]:
df = pd.read_csv('../datasets/Telco_Customer_Churn_Clean.csv')


In [120]:
#Visualizar datos preliminares del dataset

print('Dataset cargado:', df.shape)
print('\nColumnas del dataset Listo y Preparado para ML:')
print(list(df.columns))
print('\nTipo de datos:')
print(df.dtypes)

Dataset cargado: (7032, 23)

Columnas del dataset Listo y Preparado para ML:
['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn', 'TenureGroup', 'TotalServices']

Tipo de datos:
customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService          int64
MultipleLines         int64
InternetService      object
OnlineSecurity        int64
OnlineBackup          int64
DeviceProtection      int64
TechSupport           int64
StreamingTV           int64
StreamingMovies       int64
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges       

## Identificando columnas utiles

Se analizaran todas las columnas y se tomaran solo las unicas columnas que el modelo puede usar para aprender.

In [121]:
columnas_descartadas = ['customerID']

df_ml = df.drop(columns=columnas_descartadas)
print('Columnas del dataset Limpio y Listo para ML:')
print(list(df_ml.columns))
print('\nForma del dataset', df_ml.shape)

Columnas del dataset Limpio y Listo para ML:
['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn', 'TenureGroup', 'TotalServices']

Forma del dataset (7032, 22)


## Trabajando con las Variables "X" e "Y"

La variable "X" es la que estara analizando usara los features para aprender. Por otro lado tendremos la variable "Y" que sera la que el modelo tratara de predecir.

In [122]:
X = df_ml.drop(columns=['tenure', 'TotalCharges','TenureGroup' ])
y= df_ml['tenure']

print("La variable 'X' features que el modelo usara para aprender" )
print("  shape:", X.shape)
print("  Columnas:", list(X.columns))
print()
print("La variable 'Y' es la variable que el modelo intentara predecir")
print("  shape:", y.shape)



La variable 'X' features que el modelo usara para aprender
  shape: (7032, 19)
  Columnas: ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'Churn', 'TotalServices']

La variable 'Y' es la variable que el modelo intentara predecir
  shape: (7032,)


## Paso 3. Train/Test Split

In [123]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
)

print("Tamaño del split:")
print(f"  X_train: {X_train.shape} - Informacion con la que el modelo aprende")
print(f"  X_test: {X_test.shape} - Informacion con la que se evalua el modelo")
print(f"  y_train: {y_train.shape}")
print(f"  y_test: {y_test.shape}")
print()
print("Tenure promedio en train:", round(y_train.mean(), 2))
print("Tenure promedio en test:", round(y_test.mean(), 2))

Tamaño del split:
  X_train: (5625, 19) - Informacion con la que el modelo aprende
  X_test: (1407, 19) - Informacion con la que se evalua el modelo
  y_train: (5625,)
  y_test: (1407,)

Tenure promedio en train: 32.34
Tenure promedio en test: 32.77


## Paso 4. Encoding

In [124]:
print(list(df.columns))

['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn', 'TenureGroup', 'TotalServices']


In [125]:
columnas_categoricas = X.select_dtypes(include='object').columns
print('Columnas a codificar: ', columnas_categoricas)
print()

X_train_enc = pd.get_dummies(X_train, columns=columnas_categoricas, drop_first=False)
X_test_enc = pd.get_dummies(X_test, columns=columnas_categoricas, drop_first=False)

#alinear columnas
X_test_enc = X_test_enc.reindex(columns=X_train_enc.columns, fill_value=0)

#Comparacion
print("Columnas ANTES del encoding: ", X_train.shape[1])
print("Columnas DESPUES del encoding: ", X_train_enc.shape[1])
print()

print('Informacion de X_train')
print(list(X_train.columns))
print()
print("Nueva Info de las columnas:")
print(list(X_train_enc.columns))

Columnas a codificar:  Index(['gender', 'Partner', 'Dependents', 'InternetService', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')

Columnas ANTES del encoding:  19
Columnas DESPUES del encoding:  30

Informacion de X_train
['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'Churn', 'TotalServices']

Nueva Info de las columnas:
['SeniorCitizen', 'PhoneService', 'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'MonthlyCharges', 'Churn', 'TotalServices', 'gender_Female', 'gender_Male', 'Partner_No', 'Partner_Yes', 'Dependents_No', 'Dependents_Yes', 'InternetService_DSL', 'InternetService_Fiber optic', 'InternetService_No', 'Contract_Month-to-month', 'Contract_One yea

### Paso 5. Escalado de Variables Numéricas

Se utilizara **StandardScaler**, el cual transforma cada variable para que tenga una media cercana a 0 y una desviación estándar cercana a 1.

Las variables binarias generadas mediante One-Hot Encoding no serán escaladas, ya que sus valores ya se encuentran en un rango adecuado (0 y 1).

In [126]:
columnas_numericas = ['MonthlyCharges', 'TotalServices']

scaler = StandardScaler()

X_train_enc[columnas_numericas] = scaler.fit_transform(X_train_enc[columnas_numericas])
X_test_enc[columnas_numericas] = scaler.transform(X_test_enc[columnas_numericas])

print("Estadisticas de las variables numericas en X_train_enc:")
print(X_train_enc[columnas_numericas].describe().round(3))
print()
print("\nMedia de las variables numericas")
print(X_train_enc[columnas_numericas].mean().round(4))
print("\nDesviacion Standard de las variables numericas")
print(X_train_enc[columnas_numericas].std().round(4))


Estadisticas de las variables numericas en X_train_enc:
       MonthlyCharges  TotalServices
count        5625.000       5625.000
mean           -0.000          0.000
std             1.000          1.000
min            -1.539         -1.632
25%            -0.978         -1.147
50%             0.187         -0.176
75%             0.835          0.795
max             1.792          2.251


Media de las variables numericas
MonthlyCharges   -0.0
TotalServices     0.0
dtype: float64

Desviacion Standard de las variables numericas
MonthlyCharges    1.0001
TotalServices     1.0001
dtype: float64


### Interpretación

El proceso de estandarización fue aplicado correctamente a las variables numéricas del conjunto de entrenamiento, obteniendo una media cercana a **0** y una desviación estándar cercana a **1**. Posteriormente, el mismo escalador fue utilizado sobre el conjunto de prueba sin recalcular los parámetros, evitando la fuga de información (*data leakage*) y dejando los datos preparados para el entrenamiento de modelos de Machine Learning.

## Paso 6. Verificación Final

Antes de entrenar un modelo de Machine Learning es importante comprobar que el proceso de preparación de datos se realizó correctamente. En esta etapa se verificará que:

- El tamaño de los conjuntos de entrenamiento y prueba sea el esperado.
- No existan valores nulos.
- Todas las variables sean numéricas.
- Las dimensiones de los datasets sean consistentes.

In [127]:
#convertir columnas booleanas a enteros

bool_columnas = X_train_enc.select_dtypes(include='bool').columns

X_train_enc[bool_columnas] = X_train_enc[bool_columnas].astype('int64')
X_test_enc[bool_columnas] = X_test_enc[bool_columnas].astype('int64')

print("X_train_enc:", X_train_enc.shape)
print("X_test_enc:", X_test_enc.shape)
print("y_train:", y_train.shape)
print('y_test:', y_test.shape)
print()

print("\nEstadísticas de tenure en train:")
print(y_train.describe())

print("\nEstadísticas de tenure en test:")
print(y_test.describe())

print("Valores nulos en X_train", X_train_enc.isnull().sum().sum())
print("Valores nulos en X_test", X_test_enc.isnull().sum().sum())
print()

print("Tipos de datos en X_train")
print(X_train_enc.dtypes.value_counts())
print()

print("Columnas finales utilizadas por el modelo:")
print(list(X_train_enc.columns))

X_train_enc: (5625, 30)
X_test_enc: (1407, 30)
y_train: (5625,)
y_test: (1407,)


Estadísticas de tenure en train:
count    5625.000000
mean       32.335289
std        24.537144
min         1.000000
25%         9.000000
50%        29.000000
75%        55.000000
max        72.000000
Name: tenure, dtype: float64

Estadísticas de tenure en test:
count    1407.000000
mean       32.767591
std        24.583371
min         1.000000
25%         9.500000
50%        29.000000
75%        56.000000
max        72.000000
Name: tenure, dtype: float64
Valores nulos en X_train 0
Valores nulos en X_test 0

Tipos de datos en X_train
int64      28
float64     2
Name: count, dtype: int64

Columnas finales utilizadas por el modelo:
['SeniorCitizen', 'PhoneService', 'MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'MonthlyCharges', 'Churn', 'TotalServices', 'gender_Female', 'gender_Male', 'Partner_No', 'Partner_Yes', 'Dependents_No', 'Dep

### Interpretación

La verificación final confirma que el proceso de preparación de datos fue completado correctamente. Los conjuntos de entrenamiento y prueba presentan las dimensiones esperadas, no contienen valores nulos y todas las variables se encuentran en formato numérico, requisito indispensable para la mayoría de los algoritmos de Machine Learning.

Además, las variables categóricas fueron transformadas mediante One-Hot Encoding y las variables numéricas fueron estandarizadas utilizando **StandardScaler**, dejando el dataset completamente preparado para el entrenamiento y evaluación de modelos predictivos de clasificación.

### Modelo de regresión: predecir la antigüedad del cliente

vamos a predecir tenure (cuántos meses lleva un cliente) a partir de su perfil y sus servicios contratados. Es una pregunta de negocio real: ¿qué tipo de cliente, con qué combinación de servicios, tiende a ser más nuevo o más antiguo?

In [129]:
modelo_reg = LinearRegression()
modelo_reg.fit(X_train_enc, y_train)

print("Coeficientes:")

for nombre, coef in zip(X_train.columns, modelo_reg.coef_):
    print(f"{nombre} : {round(coef, 2)}")

print()
print('Intercepto:', round(modelo_reg.intercept_, 2))

Coeficientes:
gender : 2.49
SeniorCitizen : -3.61
Partner : 5.97
Dependents : 2.53
PhoneService : 4.59
MultipleLines : 1.33
InternetService : -0.86
OnlineSecurity : 1.34
OnlineBackup : 0.91
DeviceProtection : -5.05
TechSupport : -6.87
StreamingTV : 5.92
StreamingMovies : -0.41
Contract : 0.41
PaperlessBilling : -3.51
PaymentMethod : 3.51
MonthlyCharges : 0.94
Churn : -0.94
TotalServices : -2.8

Intercepto: 34.92


In [130]:
#Metricas

y_pred_reg = modelo_reg.predict(X_test_enc)

mae_reg = mean_absolute_error(y_test, y_pred_reg)
mse_reg = mean_squared_error(y_test, y_pred_reg)
rmse_reg = mse_reg ** 0.5
r2_reg = r2_score(y_test, y_pred_reg)

print("Metricas del Modelo de Regresion Lineal Multiple")
print("MAE (Error Absoluto Medio): ", round(mae_reg, 2))
print("MSE (Error Cuadratico Medio): ", round(mse_reg, 2))
print("RMSE (Raiz del error Cuadratico Medio): ", round(rmse_reg, 2))
print("R2 (Coeficiente de determinacion): ", round(r2_reg, 2))

Metricas del Modelo de Regresion Lineal Multiple
MAE (Error Absoluto Medio):  11.17
MSE (Error Cuadratico Medio):  210.17
RMSE (Raiz del error Cuadratico Medio):  14.5
R2 (Coeficiente de determinacion):  0.65


### Detección y corrección de Data Leakage

Durante el primer entrenamiento, el modelo de Regresión Lineal Múltiple obtuvo un
R² de 0.96, un resultado sospechosamente alto.

Se identificaron dos variables que producían fuga de información:

- `TotalCharges`: está relacionada aproximadamente con `MonthlyCharges × tenure`.
- `TenureGroup`: fue creada directamente a partir de la variable objetivo `tenure`.

Después de eliminar ambas variables del conjunto de predictores y volver a
entrenar el modelo, el R² disminuyó a **0.65**, con un MAE de **11.17 meses**
y un RMSE de **14.50 meses**.

Aunque el rendimiento disminuyó, estas métricas representan una evaluación más
realista de la capacidad del modelo para predecir `tenure` sin utilizar
información derivada de la variable objetivo.

### Interpretacion de las Metricas finales

El modelo consigue explicar aproximadamente un 65% de la variabilidad de **'tenure'** utilizando caracteristicas del cliente como servicos contratados, tipo de contrato, cargos mensuales, etc... segun la metrica del coeficiente de determinacion. Por otro lado, acorde al MAE las predicciones se desvian del valor real aproximadamente 11.17 meses en promedio.



### Probar el modelo con un cliente nuevo - ¿cuánto tiempo se estima que permanecerá como cliente?

In [164]:
cliente_nuevo = pd.DataFrame([{
    'customerID': 'NEW-0001',
    'gender': 'Female',
    'SeniorCitizen': 0,
    'Partner': 'No',
    'Dependents': 'No',
    'tenure': 6,
    'PhoneService': 'Yes',
    'MultipleLines': 'No',
    'InternetService': 'Fiber optic',
    'OnlineSecurity': 'No',
    'OnlineBackup': 'No',
    'DeviceProtection': 'No',
    'TechSupport': 'No',
    'StreamingTV': 'Yes',
    'StreamingMovies': 'Yes',
    'Contract': 'Month-to-month',
    'PaperlessBilling': 'Yes',
    'PaymentMethod': 'Electronic check',
    'MonthlyCharges': 90.00,
    'TotalCharges': 540.00,
    'Churn': 0,
    'TenureGroup': 'Nuevo',
    'TotalServices': 4
}])

client_regresion = cliente_nuevo.copy()
columnas_eliminar_reg = ['customerID', 'tenure', 'TotalCharges', 'TenureGroup']
client_regresion = client_regresion.drop(columns=columnas_eliminar_reg)

#one-hot encoding
client_reg_enc = pd.get_dummies(client_regresion)

#Alinear columnas con X_train
client_reg_enc = client_reg_enc.reindex(columns=X_train_enc.columns, fill_value=0)

#Comprobar que el cliente nuevo este alineado y listo para usarlo con el modelo
print('Dimensiones del Cliente Nuevo:',client_reg_enc.shape[1])
print('Dimensiones del X_train:',X_train_enc.shape[1])

#Convertir columnas booleanas a int
Columnas_bool_nuevo_cliente = client_reg_enc.select_dtypes('bool').columns
client_reg_enc[Columnas_bool_nuevo_cliente] = client_reg_enc[Columnas_bool_nuevo_cliente].astype('int')

#Standard-Scaler
client_reg_enc[columnas_numericas] = scaler.transform(client_reg_enc[columnas_numericas])

#Validacion final de nuevo cliente antes de pasarla al modelo:
print('\nTipos de datos del nuevo cliente')
print(client_reg_enc.dtypes.value_counts())
print()
print("Tiene las mismas columnas que X_train?:",
      client_reg_enc.columns.equals(X_train_enc.columns))

Dimensiones del Cliente Nuevo: 30
Dimensiones del X_train: 30

Tipos de datos del nuevo cliente
int64      28
float64     2
Name: count, dtype: int64

Tiene las mismas columnas que X_train?: True


In [165]:
#Usar el nuevo cliente con el modelo

tenure_predicho = modelo_reg.predict(client_reg_enc)

print(f"Tenure estimado: {tenure_predicho[0]:.2f} meses")


Tenure estimado: 19.14 meses


### Interpretacion del Nuevo/Desconocido cliente

Basándose únicamente en el perfil y los servicios del cliente hipotético, el modelo estima que este cliente tendría una antigüedad aproximada de 19.14 meses.